phd-tests/framework/src/serial/raw_buffer.rs

```diff
@@ -0,0 +1,141 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implements a "raw" buffer for serial console output that processes
+//! characters and newlines but ignores VT100 control characters.
+
+use std::io::{BufWriter, Write};
+
+use anyhow::{Context, Result};
+use camino::Utf8PathBuf;
+use tracing::error;
+
+use super::{Buffer, OutputWaiter};
+
+/// Wraps up the contents of a raw buffer in a single struct that can implement
+/// [`vte::Perform`]. This allows the "outer" buffer to contain a
+/// [`vte::Parser`] and call its `advance` method with a mutable reference to
+/// the buffer contents as the first argument without running afoul of mutable
+/// borrowing rules.
+struct Inner {
+    log: std::io::BufWriter<std::fs::File>,
+    line_buffer: String,
+    wait_buffer: String,
+    waiter: Option<OutputWaiter>,
+}
+
+impl Inner {
+    /// Pushes `c` to the buffer's contents and attempts to satisfy active
+    /// waits.
+    fn push_character(&mut self, c: char) {
+        if c == '\n' {
+            writeln!(self.log, "{}", &self.line_buffer).unwrap();
```

Comment:
very tiny nit: i generally prefer to avoid going through the `fmt` machinery when there's no actual formatting going on:
```suggestion
            self.log.write_all(self.line_buffer.as_bytes()).unwrap();
```
not a blocker, of course!

Reply:
I knew there had to be a better way to do this. Thanks! Fixed in 6e75a993.